pkg/api/pod/warnings.go

```diff
@@ -221,6 +224,16 @@ func warningsForPodSpecAndMeta(fieldPath *field.Path, podSpec *api.PodSpec, meta
 			}
 		}
 
+		// use of container AppArmor annotation without accompanying field
+		// Only used with pod templates; pod warnings are emitted through applyAppArmorVersionSkew instead.
+		if utilfeature.DefaultFeatureGate.Enabled(features.AppArmorFields) &&
+			fieldPath != nil && (c.SecurityContext == nil || c.SecurityContext.AppArmorProfile == nil) && !hasPodAppArmorProfile {
+			key := api.DeprecatedAppArmorAnnotationKeyPrefix + c.Name
+			if _, exists := meta.Annotations[key]; exists {
```

Comment:
nit: I think the logic is correct, but it's twisty to read. naming and isolating the checks might help?

```go
if utilfeature.DefaultFeatureGate.Enabled(features.AppArmorFields) {
  isPodTemplate := fieldPath != nil
  annotationKey := api.DeprecatedAppArmorAnnotationKeyPrefix + c.Name
  if _, hasAnnotation := meta.Annotations[annotationKey]; hasAnnotation && isPodTemplate {
    hasContainerAppArmorProfile := c.SecurityContext != nil && c.SecurityContext.AppArmorProfile != nil
    if !hasContainerAppArmorProfile && !hasPodAppArmorProfile {
      warnings = append(warnings, fmt.Sprintf(`%s: deprecated since v1.30; use the "appArmorProfile" field instead`, fieldPath.Child("metadata", "annotations").Key(annotationKey)))
    }
  }
}
```

Reply:
Done (slightly different order).